### ISSUE TITLE
About the Shape_formation_block

### ISSUE THREAD

**Issue by hanssssssss** [microsoft/Deep3DFaceReconstruction#113]
Hi !Thanks for sharing your great work ! I have no idea about the code in the function called 'Shape_formation_block',
`		face_shape = face_shape - tf.reshape(tf.reduce_mean(tf.reshape(facemodel.meanshape,[-1,3]),0),[1,1,3])
`I have read your annotation，it is used to re-center the face shape with mean shape.But I still couldn't understand ,Could you please give me more explanations?

**Comment by YuDeng** [comment]
We re-center face shapes so that for a shape with arbitrary rotation angles, when projected onto the image plane, will have similar size to a frontal face and lies in the middle area of the image. 

In our method, we use a cropped region of BFM which only contains face region. Therefore, the original center of BFM model is a little bit behind. If use the original center (without re-centering), faces with non-zero rotation angles will not fall in the center of the image.

**Comment by hanssssssss** [comment]
> We re-center face shapes so that for a shape with arbitrary rotation angles, when projected onto the image plane, will have similar size to a frontal face and lies in the middle area of the image.
> 
> In our method, we use a cropped region of BFM which only contains face region. Therefore, the original center of BFM model is a little bit behind. If use the original center (without re-centering), faces with non-zero rotation angles will not fall in the center of the image.

Thanks a lot!!!

**Comment by hanssssssss** [comment]
> We re-center face shapes so that for a shape with arbitrary rotation angles, when projected onto the image plane, will have similar size to a frontal face and lies in the middle area of the image.
> 
> In our method, we use a cropped region of BFM which only contains face region. Therefore, the original center of BFM model is a little bit behind. If use the original center (without re-centering), faces with non-zero rotation angles will not fall in the center of the image.
Hi ! Sorry for bothering you again. I also have question about the meaning of the point_buf in the BFM model.
`		self.point_buf = tf.constant(model['point_buf']) # face indices for each vertex that lies in. starts from 1. [N,8]
`
Does it mean the adjacent face index for each vertex?

**Comment by YuDeng** [comment]
> > We re-center face shapes so that for a shape with arbitrary rotation angles, when projected onto the image plane, will have similar size to a frontal face and lies in the middle area of the image.
> > In our method, we use a cropped region of BFM which only contains face region. Therefore, the original center of BFM model is a little bit behind. If use the original center (without re-centering), faces with non-zero rotation angles will not fall in the center of the image.
> 
> Hi ! Sorry for bothering you again. I also have question about the meaning of the point_buf in the BFM model.
> ` self.point_buf = tf.constant(model['point_buf']) # face indices for each vertex that lies in. starts from 1. [N,8] `
> Does it mean the adjacent face index for each vertex?

Yes it is.

**Comment by hanssssssss** [comment]
> > > We re-center face shapes so that for a shape with arbitrary rotation angles, when projected onto the image plane, will have similar size to a frontal face and lies in the middle area of the image.
> > > In our method, we use a cropped region of BFM which only contains face region. Therefore, the original center of BFM model is a little bit behind. If use the original center (without re-centering), faces with non-zero rotation angles will not fall in the center of the image.
> > 
> > 
> > Hi ! Sorry for bothering you again. I also have question about the meaning of the point_buf in the BFM model.
> > `self.point_buf = tf.constant(model['point_buf']) # face indices for each vertex that lies in. starts from 1. [N,8]`
> > Does it mean the adjacent face index for each vertex?
> 
> Yes it is.

Thank you so much. I appreciate your help.

**Comment by hanssssssss** [comment]
> > > We re-center face shapes so that for a shape with arbitrary rotation angles, when projected onto the image plane, will have similar size to a frontal face and lies in the middle area of the image.
> > > In our method, we use a cropped region of BFM which only contains face region. Therefore, the original center of BFM model is a little bit behind. If use the original center (without re-centering), faces with non-zero rotation angles will not fall in the center of the image.
> > 
> > 
> > Hi ! Sorry for bothering you again. I also have question about the meaning of the point_buf in the BFM model.
> > `self.point_buf = tf.constant(model['point_buf']) # face indices for each vertex that lies in. starts from 1. [N,8]`
> > Does it mean the adjacent face index for each vertex?
> 
> Yes it is.

Hi! Sorry for bothering you again, I have no idea about the perspective projection of the 3D face. I have searched many information on the internet and found that the projection matrix is a 4*4 matrix like this.But I found that the p_matrix in your work is a 3*3 matrix.
![图片](https://user-images.githubusercontent.com/57829884/108485340-c5fa9180-72d7-11eb-83f5-eb6f18257aea.png)
Could you kindly give me some information about the prospective projection in your work ?
Thanks again

**Comment by YuDeng** [comment]
Hi, the 3x3 projection matrix we use and the 4x4 matrix are doing the same thing but in different mathematical forms. You may check https://en.wikipedia.org/wiki/Pinhole_camera_model to get a rough understanding about pinhole camera model.

**Comment by hanssssssss** [comment]
Thanks for your help and your patience!!^_^

**Comment by hanssssssss** [comment]
> > > We re-center face shapes so that for a shape with arbitrary rotation angles, when projected onto the image plane, will have similar size to a frontal face and lies in the middle area of the image.
> > > In our method, we use a cropped region of BFM which only contains face region. Therefore, the original center of BFM model is a little bit behind. If use the original center (without re-centering), faces with non-zero rotation angles will not fall in the center of the image.
> > 
> > 
> > Hi ! Sorry for bothering you again. I also have question about the meaning of the point_buf in the BFM model.
> > `self.point_buf = tf.constant(model['point_buf']) # face indices for each vertex that lies in. starts from 1. [N,8]`
> > Does it mean the adjacent face index for each vertex?
> 
> Yes it is.

Hi! Sorry for bothering you again. I could not understanding the the meaning of the number '102' in the pre_processing file. In function `resize_n_crop_img(img,lm,t,s,target_size = 224.)`
`	w = (w0/s*102).astype(np.int32)
`
And I also couldn't figure out the experssion `left = (w/2 - target_size/2 + float((t[0] - w0/2)*102/s)).astype(np.int32)` in the same function. I konw the t[0] and t[1] is the translation of the x and y, and s is the sacle factors from 3d-landmark to 2d-landmark. But I dont konw the reason why should add `(t[0] - w0/2)*102/s` and `(h0/2 - t[1])*102/s)` while caculating the `left` and `up  `and when caculating the `right `and `below `doesn't need to add these.

Thanks again.

**Comment by YuDeng** [comment]
For the first question, 102 is just a manually set scalar to adjust the face size in the cropped image.

For the second question, we already know the target size of the cropped image, which is 224 in our setting. Therefore, once we get left and top in the original image, we directly know the right and below, which equal to left + 224 and top + 224 respectively.

**Comment by hanssssssss** [comment]
> For the first question, 102 is just a manually set scalar to adjust the face size in the cropped image.
> 
> For the second question, we already know the target size of the cropped image, which is 224 in our setting. Therefore, once we get left and top in the original image, we directly know the right and below, which equal to left + 224 and top + 224 respectively.

Thanks for your reply, I could understand the meanging of '102'.But I don't know why you use `(t[0] - w0/2)*102/s` when caculating the 'left'.In my points of view, the left is`(w/2 -target_size/2)* 102/s - t[0]`,because from the least square problem we knpw that 'lm3d * R + t = lm2d'. I can not understand why is `(w/2 - target_size/2 + float((t[0] - w0/2)*102/s)).`
Thanks again.

**Comment by YuDeng** [comment]
If we ignore the rotation, the least square problem is solving s*lm3D + t = lm2D(Original image space).
Therefore, in order to put the face in the center of the original image, we should move the image center by t - w0/2 (because the image center of the original image is w0/2 instead of 0). However, for the resized image to be cropped, it is scaled by 102/s. So to move the face to the center in the resized image space, we should multiply t - w0/2 with scale factor 102/s, and that deduces the equation (w/2 - target_size/2 + float((t[0] - w0/2)*102/s)).

**Comment by hanssssssss** [comment]
> If we ignore the rotation, the least square problem is solving s*lm3D + t = lm2D(Original image space).
> Therefore, in order to put the 2D face in the center of the original image, we should move the image center by t - w0/2 (because the image center of the original image is w0/2 instead of 0, but the center of lm3D is zero). However, for the resized image to be cropped, it is scaled by 102/s. So to move the face to the center in the resized image space, we should multiply t - w0/2 with scale factor 102/s, and that deduces the equation (w/2 - target_size/2 + float((t[0] - w0/2)*102/s)).

Thanks!!!! With your guidance, I understanded that reason just now.

**Comment by hanssssssss** [comment]
> Hi, the 3x3 projection matrix we use and the 4x4 matrix are doing the same thing but in different mathematical forms. You may check https://en.wikipedia.org/wiki/Pinhole_camera_model to get a rough understanding about pinhole camera model.

Hi ! Sorry for bothering you again. I have some problems about the caculation of landmark loss. According to the caculation in the function `Projection_block`, we can get a 2d-landmark. But I found that the  result of the `Pinhole_camera_model ` is a reversed image like the picture below. The onlt idea I can come up with is to transpose the image befor sending to the face-net(and I got the landmark loss 7045.1685 with single image ). Do you have other ideas?



the red landamarks is from the `face-net`
the blue landmarks is from the `projection block`
![image](https://user-images.githubusercontent.com/57829884/109281094-d7ebaf80-7856-11eb-8421-92a62f65c593.png)

I do transpose for the image befor sending to the face-net. But the nhumber of landmark loss is too large.Do you have any idea？


Thanks again.
